# S.A.B.A.T - Semi-Automatic BLAST Annotation Toolkit

SABAT is toolkit for identifying gene homologs in other species using a gene coding sequence (CDS) and BLAST.

Analysis is perfomed in 5 steps as follows:
1. BLAST your gene CDS against a genome of interest. (blastn/dc-megablast)
2. Convert the BLAST output to a BED file. (SABAT blast2bed)
3. Visualise your BED file (BLAST results) with the genome of interest. (IGV)
4. Select exons that make up your genes.
5. Reconstruct the homelogs CDS and protein sequence. (SABAT assemble-gene)

# Installation

TBA

# Commands

## blast2bed

### parameters
|Parameter|Description|
|---|---|
| -i, --input | BLAST ouput in tabular format (outfmt6) |
| -e, --exons | Expected number of exon in the gene. Default: 1 |
| -c, --coverage | A float that roportion of gene CDS that must be covered by BLAST hits for a locus to be defined (0.0-1.0). Default: 1.1 |
| -l, --locus_size | Expected size of the gene locus in base pairs. Default: 1000|
| -r, --refseq | Boolean values that indicates whether the BLASTDB provided uses a refseq sequence (RefSeq headers are different from the standard headers in a BLASTDB). Default: False |

The BED file generated is directed to the stdout.

### Usage

    python3 __main__.py blast2bed \
    --input results.blastn \
    --exons 3 \
    --coverage 0.95 \
    --locus_size 5000 \
    --refseq False \
    > output.bed

## assemble-gene

### Parameters
|Parameter |Description |
|---|---|
| -i, --input | Path to a Bed file (generated by blast2bed)|
| -db, --database | Path to BLASTDB used to generate the blast output used by blast2bed|
| -e, --exons | Path to a txt file containing the gene exons, in order, with 1 exon per line |

### Usage

Dummy exon file:

        exon_1
        exon_2
        exon_3

Command:

    python3 __main__.py assemble-gene \
    -i output.bed
    -db path/to/blast/db/
    -e exons.txt
    > genes.fasta


## Assumptions

1. A gene's first exon starts with start codon (ATG).
2. A gene's last exon ends with a stop codon (TGA, TTA or TAG).
3. A gene's intermediate exons donot have to start with a start codon (ATG)
4. If a gene's intermediate exon contains a stop codon that is the end of the exon.
5. dc-megaBLAST can more readily identify the first exon (at least for nlr/kinase) so the first exon must start with ATG
6. dc-megaBLAST less readily identifies the end of the last exon (including the stop codon) so the last exon can be extended (-flank) to include additional codons
7. The longest exon is the correct exon (providing all other criteria are satisfied) 


### Roadmap

- edit and unedited outputs for assemble-gene (will need to output results as file rather than to std out)
- chromosome headers should not contain the character |